File: prog.py
#!/usr/bin/env python3
# coding=utf-8
import sys
from spidriver import SPIDriver

if __name__ == '__main__':
    # Ensure the script is called with the correct arguments
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <device> <file>")
        sys.exit(1)

    device = sys.argv[1]
    bin_file = sys.argv[2]

    try:
        # Open SPI driver
        s = SPIDriver(device)

        # Read the binary file
        with open(bin_file, "rb") as f:
            data = f.read()

        # Write the binary data to the SPI device
        s.write(list(data))
        print(f"Successfully wrote {len(data)} bytes from {bin_file} to {device}")

        # Pull chip select high (this step assumes that the SPI driver has a way to manipulate CS)
        s.unsel()  # This method needs to exist or be added in your SPIDriver class

        # Write 49 dummy bytes (0x00)
        dummy_bytes = [0] * 49
        s.write(dummy_bytes)
        print("Successfully wrote 49 dummy bytes to the SPI device")
        s.sel()

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

File: prog.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "spidriver/c/common/spidriver.h"

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <device> <file>\n", argv[0]);
        return 1;
    }

    const char *device = argv[1];
    const char *bin_file = argv[2];
    
    SPIDriver sd;
    FILE *fp;
    char *buffer;
    long file_size;
    
    // Initialize SPI driver
    spi_connect(&sd, device);
    if (!sd.connected) {
        fprintf(stderr, "Error: Could not connect to SPI device %s\n", device);
        return 1;
    }
    
    // Open binary file
    fp = fopen(bin_file, "rb");
    if (!fp) {
        fprintf(stderr, "Error: Could not open file %s\n", bin_file);
        return 1;
    }
    
    // Get file size
    fseek(fp, 0, SEEK_END);
    file_size = ftell(fp);
    rewind(fp);
    
    // Allocate memory for file content
    buffer = (char*)malloc(file_size);
    if (!buffer) {
        fprintf(stderr, "Error: Memory allocation failed\n");
        fclose(fp);
        return 1;
    }
    
    // Read file content
    if (fread(buffer, 1, file_size, fp) != file_size) {
        fprintf(stderr, "Error: Could not read the entire file\n");
        free(buffer);
        fclose(fp);
        return 1;
    }
    fclose(fp);
    
    // Select chip
    spi_sel(&sd);
    
    // Write the binary data to the SPI device
    spi_write(&sd, buffer, file_size);
    printf("Successfully wrote %ld bytes from %s to %s\n", file_size, bin_file, device);
    
    // Unselect chip
    spi_unsel(&sd);
    
    // Write 49 dummy bytes (0x00)
    char dummy_bytes[49] = {0};
    spi_write(&sd, dummy_bytes, 49);
    printf("Successfully wrote 49 dummy bytes to the SPI device\n");
    
    // Select chip again (as in Python script)
    spi_sel(&sd);
    
    // Clean up
    free(buffer);
    
    return 0;
}